internal/asm/buffer.go

```diff
@@ -0,0 +1,220 @@
+// the garbage collector and therefore must be released *manually* by calling
+// their Unmap method to prevent memory leaks.
+//
+// The zero value is a valid, empty code segment, equivalent to being
+// constructed by calling NewCodeSegment(nil).
+type CodeSegment struct {
+	code []byte
+	size int
+}
+
+// NewCodeSegment constructs a CodeSegment value from a byte slice.
+//
+// No validation is made that the byte slice is a memory mapped region which can
+// be unmapped on Close.
+func NewCodeSegment(code []byte) *CodeSegment {
+	return &CodeSegment{code: code, size: len(code)}
+}
+
+// Map allocates a memory mapping of the given size to the code segment.
+//
+// Note that programs only need to use this method to initialize the code
+// segment to a specific content (e.g. when loading pre-compiled code from a
+// file), otherwise the backing memory mapping is allocated on demand when code
+// is written to the code segment via Buffers returned by calls to Next.
+//
+// The method errors is the segment is already backed by a memory mapping.
+func (seg *CodeSegment) Map(size int) error {
+	if seg.code != nil {
+		return fmt.Errorf("code segment already initialized to memory mapping of size %d", len(seg.code))
+	}
+	b, err := platform.MmapCodeSegment(size)
+	if err != nil {
+		return err
+	}
+	seg.code = b
+	seg.size = size
+	return nil
+}
+
+// Close unmaps the underlying memory region held by the code segment, clearing
+// its state back to an empty code segment.
+//
+// The value is still usable after unmapping its memory, a new memory area can
+// be allocated by calling Map or writing to the segment.
+func (seg *CodeSegment) Unmap() error {
+	if seg.code != nil {
+		if err := platform.MunmapCodeSegment(seg.code[:cap(seg.code)]); err != nil {
+			return err
+		}
+		seg.code = nil
+		seg.size = 0
+	}
+	return nil
+}
+
+// Addr returns the address of the beginning of the code segment as a uintptr.
+func (seg *CodeSegment) Addr() uintptr {
+	if len(seg.code) > 0 {
+		return uintptr(unsafe.Pointer(&seg.code[0]))
+	}
+	return 0
+}
+
+// Size returns the size of code segment, which is less or equal to the length
+// of the byte slice returned by Len or Bytes.
+func (seg *CodeSegment) Size() uintptr {
+	return uintptr(seg.size)
+}
+
+// Len returns the length of the byte slice referencing the memory mapping of
+// the code segment.
+func (seg *CodeSegment) Len() int {
+	return len(seg.code)
+}
+
+// Bytes returns a byte slice to the memory mapping of the code segment.
+//
+// The returned slice remains valid until more bytes are written to a buffer
+// of the code segment, or Unmap is called.
+func (seg *CodeSegment) Bytes() []byte {
+	return seg.code
+}
+
+// Next returns a buffer pointed at the end of the code segment to support
+// writing more code instructions to it.
+//
+// Buffers are passed by value, but they hold a reference to the code segment
+// that they were created from.
+func (seg *CodeSegment) Next() Buffer {
```

Comment:
this method name (`Next`) was pretty opaque to me until I had read the comment on `CodeSegment`. Is this a common name, or may I suggest to think of a better name? I can't really think of a good one :D maybe `StartSection`? `BeginAligned`? `NextAligned`?

Reply:
See the name change in https://github.com/tetratelabs/wazero/pull/1481/commits/202b640c728e8576dc95f1cd261784baa8344f03